crates/build-rs/src/output.rs

```diff
@@ -0,0 +1,408 @@
+/// closely as possible in order to avoid typos, missing check_cfg, stale cfgs,
+/// and other mistakes.
+///
+/// [`unexpected_cfgs`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unexpected-cfgs
+#[doc = msrv!("1.80")]
+#[track_caller]
+pub fn rustc_check_cfgs(keys: &[&str]) {
+    if keys.is_empty() {
+        return;
+    }
+    for key in keys {
+        if !is_ident(key) {
+            panic!("cannot emit rustc-check-cfg: invalid key {key:?}");
+        }
+    }
+
+    if allow_use::check_cfg() {
+        let mut directive = keys[0].to_string();
+        for key in &keys[1..] {
+            write!(directive, ", {key}").expect("writing to string should be infallible");
+        }
+        emit("rustc-check-cfg", format_args!("cfg({directive})"));
+    }
+}
+
+/// Add to the list of expected config names that is used when checking the
+/// *reachable* cfg expressions with the [`unexpected_cfgs`] lint.
+///
+/// This form is for keys with expected values, such as `cfg(name = "value")`.
+///
+/// It is recommended to group the `rustc_check_cfg` and `rustc_cfg` calls as
+/// closely as possible in order to avoid typos, missing check_cfg, stale cfgs,
+/// and other mistakes.
+///
+/// [`unexpected_cfgs`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unexpected-cfgs
+#[doc = msrv!("1.80")]
+#[track_caller]
+pub fn rustc_check_cfg_values(key: &str, values: &[&str]) {
+    if !is_ident(key) {
+        panic!("cannot emit rustc-check-cfg: invalid key {key:?}");
+    }
+    if values.is_empty() {
+        rustc_check_cfgs(&[key]);
+        return;
+    }
+
+    if allow_use::check_cfg() {
+        let mut directive = format!("\"{}\"", values[0].escape_default());
+        for value in &values[1..] {
+            write!(directive, ", \"{}\"", value.escape_default())
+                .expect("writing to string should be infallible");
+        }
+        emit(
+            "rustc-check-cfg",
+            format_args!("cfg({key}, values({directive}))"),
+        );
+    }
+}
+
+/// The `rustc-env` instruction tells Cargo to set the given environment variable
+/// when compiling the package. The value can be then retrieved by the
+/// [`env!` macro][env!] in the compiled crate. This is useful for embedding
+/// additional metadata in crate’s code, such as the hash of git HEAD or the
+/// unique identifier of a continuous integration server.
+///
+/// See also the [environment variables automatically included by Cargo][cargo-env].
+///
+/// [cargo-env]: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates
+#[track_caller]
+pub fn rustc_env(key: &str, value: &str) {
+    if key.contains(['=', '\n']) {
+        panic!("cannot emit rustc-env: invalid key {key:?}");
+    }
+    if value.contains('\n') {
+        panic!("cannot emit rustc-env: invalid value {value:?}");
+    }
+    emit("rustc-env", format_args!("{key}={value}"));
```

Comment:
Need to check/test that the various directives apply quoting/escaping correctly

Reply:
While code feedback is good, we should probably decouple it from this PR and handle it in follow ups.